### PR TITLE
Refactor to add share modal feature

### PR DESCRIPTION
**Pull request by ragurney**

### CODE REVIEW

File: test/flows/share_image_test.rb [hunk 1]
```diff
@@ -14,23 +14,54 @@ class ShareImageTest < FlowTestCase
...
+    images_show_page.share_image do |modal|
+      assert_equal happening_url, modal.image_url
+      modal.email_address.set('valid@email.com')
+      modal.message.set('Message!')
+      modal.share
+    end
+    assert_equal 'Email successfully sent!', images_show_page.flash_message(:success)
+  end
+
+  test 'share image with modal with inline errors' do
```
Comment (rmacklin) [inline test/flows/share_image_test.rb]: I think if you want to split this out into its own test that should be a separate commit. Doing so would keep this commit focused purely on swapping out the share page for the modal, and the flow test diff should contain as few changes as possible (just the ones related to the share page -> modal change, and any additional modal-specific test cases at the end).

File: test/flows/share_image_test.rb [hunk 2]
```diff
@@ -14,23 +14,54 @@ class ShareImageTest < FlowTestCase
...
+      assert_equal 'Please review the problems below:', images_show_page.flash_message(:danger)
+      assert_equal "can't be blank", modal.email_address.error_message
 
-    share_new_image_page.email_address.set('invalid_email')
-    share_new_image_page = share_new_image_page.share_image!.as_a(PageObjects::Images::ShareNewPage)
-    assert_equal 'Please review the problems below:', share_new_image_page.flash_message(:danger)
-    assert_equal 'is invalid', share_new_image_page.email_address.error_message
+      modal.email_address.set('invalid_email')
+      modal.share
+      assert_equal 'Please review the problems below:', images_show_page.flash_message(:danger)
```
Comment (rmacklin) [inline test/flows/share_image_test.rb]: This isn't a flash message on the show page. It's an alert inside the modal. If this assertion is passing, that indicates a flaw in the page object interface.

File: test/flows/share_image_test.rb [hunk 3]
```diff
@@ -42,13 +73,15 @@ class ShareImageTest < FlowTestCase
       image.url == happening_url
     end
 
-    share_new_image_page = image_to_share.open_share_page!
-    created_image.destroy!
+    image_to_share.share_image do |modal|
+      created_image.destroy
```
Comment (rmacklin) [inline test/flows/share_image_test.rb]: This doesn't need to be inside the block.

File: test/flows/share_image_test.rb [hunk 4]
```diff
@@ -42,13 +73,15 @@ class ShareImageTest < FlowTestCase
...
+      modal.email_address.set('valid@email.com')
+      modal.message.set('Some message')
+      modal.share
+    end
```
Comment (rmacklin) [inline test/flows/share_image_test.rb]: This is still an issue: https://github.com/AppFolioOnboarding/image-sharer-ragurney/pull/42#discussion_r80969785. The 404 test needs to use the `open_share_modal` method (which is currently unused).

File: test/page_objects/images/share_modal.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+module PageObjects
+  module Images
+    class ShareModal < AePageObjects::Element
```
Comment (rmacklin) [inline test/page_objects/images/share_modal.rb]: This page object isn't brand new. It's just a modification of the [`ShareNewPage`](https://github.com/AppFolioOnboarding/image-sharer-ragurney/blob/8fdb5978ee33f3b86d71ffe42e708f2008736902/test/page_objects/images/share_new_page.rb). The history should reflect that (and the `ShareNewPage` should go away).

File: app/assets/javascripts/images/set_flash_message.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class SetFlashMessage {
```
Comment (ragurney) [inline app/assets/javascripts/images/set_flash_message.js]: 320f234

File: app/assets/javascripts/images/modal_driver.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import ShareImageModal from './share_image_modal';
+import SetFlashMessage from './set_flash_message';
+
+const flashMessageGenerator = new SetFlashMessage();
+const shareImageModal = new ShareImageModal('#share_image_modal', flashMessageGenerator);
+
+shareImageModal.attachEventHandlers();
+
```
Comment (ragurney) [inline app/assets/javascripts/images/modal_driver.js]: 320f234

File: app/assets/javascripts/images/share_image_modal.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    });
+
+    this.$modal.on('hide.bs.modal', event => {
+      this.clearForm();
+    });
+
+    this.$modal.on('ajax:success', (e, xhr) => {
+      this.$modal.modal('hide');
+      this.flashMessageGenerator.generateFlashMessage('success', 'Email successfully sent!');
+    }).on('ajax:error', (e, xhr) => {
```
Comment (ragurney) [inline app/assets/javascripts/images/share_image_modal.js]: 23b67e6

File: app/assets/javascripts/images/share_image_modal.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        this.$shareForm.html(xhr.responseJSON.share_form_html);
+      }
+      else if (xhr.status === 404) {
+        this.$modal.modal('hide');
+        window.location.replace('/');
+      }
+      else {
+        this.$modal.modal('hide');
+        this.flashMessageGenerator.generateFlashMessage('danger',
+          'Something went terribly wrong. Please try again.')
```
Comment (ragurney) [inline app/assets/javascripts/images/share_image_modal.js]: e9a1af5

File: app/views/images/_share_form.html.erb [hunk 10]
```diff
@@ -1,23 +1,20 @@
-<h1 class="text-xs-center">Share Your Image</h1>
+<%= simple_form_for share_form, url: image.nil? ? '' : share_send_image_path(image), remote: true do |f| %>
+    <div class="modal-body">
```
Comment (ragurney) [inline app/views/images/_share_form.html.erb]: 5d1194f

File: test/flows/share_image_test.rb [hunk 11]
```diff
@@ -1,7 +1,7 @@
 require 'flow_test_helper'
 
 class ShareImageTest < FlowTestCase
-  test 'share image' do
+  test 'share image with modal with no errors' do
```
Comment (ragurney) [inline test/flows/share_image_test.rb]: 18a82fd

File: test/flows/share_image_test.rb [hunk 12]
```diff
@@ -14,23 +14,54 @@ class ShareImageTest < FlowTestCase
     image_to_share = images_index_page.images.find do |image|
       image.url == mind_blown_url
     end
+    image_to_share.share_image do |modal|
```
Comment (ragurney) [inline test/flows/share_image_test.rb]: 18a82f

File: test/flows/share_image_test.rb (hunk 2 above)
Comment (ragurney) [inline test/flows/share_image_test.rb]: b368811

File: test/flows/share_image_test.rb (hunk 3 above)
Comment (ragurney) [inline test/flows/share_image_test.rb]: abedc9c

File: test/flows/share_image_test.rb (hunk 4 above)
Comment (ragurney) [inline test/flows/share_image_test.rb]: c21f26f

**Comment by ragurney** [comment]
I've added some fixes to address your comments. The ones with '👍 ' are commit history fixes which I will take care of once all of the other issues with this branch are resolved.

File: test/page_objects/images/share_modal.rb [hunk 13]
```diff
@@ -6,6 +6,10 @@ class ShareModal < AePageObjects::Element
         element :message
       end
 
+      def alert(message_type)
```
Comment (rmacklin) [inline test/page_objects/images/share_modal.rb]: Same comments: I don't think `message_type` is necessary, and this could use a better name like `form_error_message` or something.

File: test/page_objects/images/share_new_page.rb [hunk 14]
```diff
@@ -2,7 +2,7 @@ module PageObjects
   module Images
     class ShareNewPage < PageObjects::Document
       path :share_new_image
-      path :share_image # If create fails
+      path :share # If create fails
```
Comment (rmacklin) [inline test/page_objects/images/share_new_page.rb]: I don't believe you meant to change this. Perhaps a find & replace mistake? Try to remember to look over your diffs before you commit.

File: test/controllers/images_controller_test.rb [hunk 15]
```diff
@@ -172,10 +172,10 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     end
     assert_response :unprocessable_entity
 
-    response_form_html_string = response.parsed_body['share_form_html']
+    response_form_html_string = Nokogiri::HTML.parse(response.parsed_body['share_form_html']).inner_html
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Glad you found a fix to remove `&#39;` from the expected value :)

But this made me think you might be able to use `assert_select` after all, if you manually construct an `HTML::Document` [like `assert_select_email` does](https://github.com/rails/rails/blob/8c81c9c423c0b463637976d5f00df760e2e733e1/actionpack/lib/action_dispatch/testing/assertions/selector.rb#L416-L417). (You may need to first pass the input to `CGI.unescapeHTML` [like `assert_select_encoded` does](https://github.com/rails/rails/blob/8c81c9c423c0b463637976d5f00df760e2e733e1/actionpack/lib/action_dispatch/testing/assertions/selector.rb#L382).)

File: app/views/images/_image_card.html.erb [hunk 16]
```diff
@@ -17,9 +17,9 @@
                      share_path: share_image_path(image),
                      image_url: image.url
                    } %>
-      <%= link_to 'Delete', image_path(image), method: :delete,
-                  class: "js-delete-image btn btn-danger",
-                  'data-confirm': 'Are you sure you want to delete this image?' %>
+    <%= link_to 'Delete', image_path(image), method: :delete,
+                class: "js-delete-image btn btn-danger",
+                'data-confirm': 'Are you sure you want to delete this image?' %>
```
Comment (rmacklin) [inline app/views/images/_image_card.html.erb]: I'm not sure why this was changed while properly indenting the `share_form`. It seems like this was properly indented before anyway.

File: app/views/images/_image_card.html.erb [hunk 17]
```diff
@@ -9,18 +9,14 @@
...
-                       data: {
-                         toggle: 'modal',
-                         target: '#share_image_modal',
-                         share_path: share_send_image_path(image),
-                         image_url: image.url
-                       } %>
-      <% else %>
-        <%= link_to 'Share', share_new_image_path(image), class: "js-share-image btn btn-info" %>
-      <% end %>
+    <%= button_tag 'Share',
```
Comment (rmacklin) [inline app/views/images/_image_card.html.erb]: This should be indented since it's inside the `<div class="card-block">` element.

File: app/assets/javascripts/images/share_image_modal.js [hunk 18]
```diff
@@ -40,8 +40,7 @@ class ShareImageModal {
         window.location.replace('/');
       }
       else {
-        this.$modal.modal('hide');
-        setFlashMessage('danger', 'Something went terribly wrong. Please try again.')
+        alert('Please try again.')
```
Comment (rmacklin) [inline app/assets/javascripts/images/share_image_modal.js]: What happened to "Something went terribly wrong."? :)

File: app/views/images/_image_card.html.erb (hunk 16 above)
Comment (ragurney) [inline app/views/images/_image_card.html.erb]: To me it wasn't lined up properly with the button_tag. I forgot to fix it's indentation after I removed the feature toggle. Up to this point, it was indented as if it was still inside the else clause.

File: test/page_objects/images/share_modal.rb (hunk 13 above)
Comment (ragurney) [inline test/page_objects/images/share_modal.rb]: 67e00dd

File: test/page_objects/images/share_new_page.rb (hunk 14 above)
Comment (ragurney) [inline test/page_objects/images/share_new_page.rb]: ecd6e8f

File: app/views/images/_image_card.html.erb (hunk 16 above)
Comment (ragurney) [inline app/views/images/_image_card.html.erb]: 21631af

File: app/views/images/_image_card.html.erb (hunk 17 above)
Comment (ragurney) [inline app/views/images/_image_card.html.erb]: 21631af

File: app/assets/javascripts/images/share_image_modal.js (hunk 18 above)
Comment (ragurney) [inline app/assets/javascripts/images/share_image_modal.js]: 9f9fb1b

File: test/page_objects/images/share_modal.rb (hunk 13 above)
Comment (rmacklin) [inline test/page_objects/images/share_modal.rb]: That commit only addresses the second comment.

File: test/page_objects/images/share_new_page.rb (hunk 14 above)
Comment (rmacklin) [inline test/page_objects/images/share_new_page.rb]: That commit is set to `fixup` the wrong commit. It was incorrectly changed in 18a82fda86c71e00750ae264ecce1136c9943b4e, not 07b0ab248d0ac0b834827f8284d90d9db6328716

File: app/views/images/_image_card.html.erb (hunk 16 above)
Comment (rmacklin) [inline app/views/images/_image_card.html.erb]: I don't think that will squash super cleanly, but the conflicts should be easy to resolve.

File: test/page_objects/images/share_new_page.rb (hunk 14 above)
Comment (ragurney) [inline test/page_objects/images/share_new_page.rb]: 25f991a

File: test/page_objects/images/share_modal.rb (hunk 13 above)
Comment (ragurney) [inline test/page_objects/images/share_modal.rb]: 25f991a

File: test/page_objects/images/share_new_page.rb (hunk 14 above)
Comment (ragurney) [inline test/page_objects/images/share_new_page.rb]: 2ff8799

File: test/flows/share_image_test.rb (hunk 1 above)
Comment (ragurney) [inline test/flows/share_image_test.rb]: 5a19882

**Comment by ragurney** [comment]
Commit history issues mentioned above have now been fixed (see inline comment responses for details). Only comment left unaddressed is [this one](https://github.com/AppFolioOnboarding/image-sharer-ragurney/pull/50#discussion_r82147307).

File: test/page_objects/images/show_page.rb [hunk 19]
```diff
@@ -9,6 +9,10 @@ class ShowPage < PageObjects::Document
 
       delegate :delete, :delete_and_confirm!, :open_share_page!, :share, to: :image
 
+      def form_error_message
+        node.find('.alert.alert-danger').text
+      end
```
Comment (rmacklin) [inline test/page_objects/images/show_page.rb]: Why is this on the show page?

Comment (ragurney) [inline test/page_objects/images/show_page.rb]: I think I got carried away with the form_error_message refactor 😅 . It's now removed eec6e3b
